warehouse/macaroons/services.py

```diff
@@ -0,0 +1,120 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import datetime
+import json
+import uuid
+
+import pymacaroons
+
+from sqlalchemy.orm import joinedload
+from sqlalchemy.orm.exc import NoResultFound
+from zope.interface import implementer
+
+from warehouse.accounts.models import User
+from warehouse.macaroons.caveats import Verifier
+from warehouse.macaroons.interfaces import IMacaroonService
+from warehouse.macaroons.models import Macaroon
+
+
+class InvalidMacaroon(Exception):
+    ...
+
+
+@implementer(IMacaroonService)
+class DatabaseMacaroonService:
+    def __init__(self, db_session):
+        self.db = db_session
+
+    def find_macaroon(self, macaroon_id):
+        """
+        Returns a macaroon model from the DB by its identifier.
+        Returns None if no macaroon has the given ID.
+        """
+        try:
+            dm = (
+                self.db.query(Macaroon)
+                .options(joinedload("user"))
+                .filter(Macaroon.id == uuid.UUID(macaroon_id))
+                .one()
+            )
+        except NoResultFound:
+            return None
+
+        return dm
+
+    def find_userid(self, raw_macaroon):
+        """
+        Returns the id of the user associated with the given raw (serialized)
+        macaroon.
+        """
+        if raw_macaroon is None:
+            return None
+        m = pymacaroons.Macaroon.deserialize(raw_macaroon)
+        dm = self.find_macaroon(m.identifier.decode())
+
+        if dm is None:
+            return None
+
+        return dm.user.id
+
+    def verify(self, raw_macaroon, context, principals, permission):
+        """
+        Returns True if the given raw (serialized) macaroon is
+        valid for the context, principals, and requested permission.
+
+        Raises InvalidMacaroon if the macaroon is not valid.
+        """
+        m = pymacaroons.Macaroon.deserialize(raw_macaroon)
+        dm = self.find_macaroon(m.identifier.decode())
+
+        if dm is None:
+            raise InvalidMacaroon
+
+        verifier = Verifier(m, context, principals, permission)
+        if verifier.verify(dm.key):
+            dm.last_used = datetime.datetime.now()
```

Comment:
This will be saved because there is a call to commit the DB session somewhere in the request lifecycle?

Reply:
Yes, that's correct. We use the same style for some changes to WebAuthn models as well (IIRC).

Reply:
For completeness sake: We're using pyramid_tm to tie the database transaction to the lifecycle of the HTTP request.